docs/src/explanations/dpi.md

```diff
@@ -0,0 +1,143 @@
+  * The output argument must be the last argument in DPI fuction.
+* Clocked vs. Unclocked:
+  * Clocked: The function call is evaluated at the associated clock's positive edge. By default, Chisel uses the current module's clock. For custom clocks, use `withClocked(clock) {..}`. If the function has a return value, it will be available in the next clock cycle.
+  * Unclocked: The function call is evaluated immediately.
+
+## Example: Adding Two Numbers
+Here's an example of a DPI function that calculates the sum of two numbers:
+
+```c++
+extern "C" void add(int lhs, int rhs, int* result)
+{
+    *result = lhs + rhs;
+}
+```
+
+```scala
+object Add extends DPINonVoidFunctionImport[UInt] {
+  override val functionName = "add"
+  override val ret = UInt(32.W)
+  override val clocked = false
+  override val inputNames = Some(Seq("lhs", "rhs"))
+  override val outputName = Some("result")
+  final def apply(lhs: UInt, rhs: UInt): UInt = super.call(lhs, rhs)
+}
+
+val result = Add(2.U(32.W), 3.U(32.W))
+```
+
+Explanation:
+
+* `Add` inherits from `DPINonVoidFunctionImport[UInt]` because it returns a 32-bit unsigned integer.
+* `ret` specifies the return type.
+* `clocked` indicates that this is a clocked function call.
+* `inputNames` and `outputName` provide optional names for the function's arguments and return value (these are just for Verilog readability).
+
+## Example: Sum of an array
+Chisel vectors are converted into SystemVerilog open arrays when used with DPI. Since memory layout can vary between simulators, it's recommended to use `svSize` and `svGetBitArrElemVecVal` to access array elements.
+
+```c++
+extern "C" void sum(const svOpenArrayHandle array, int* result) {
+  // Get a length of the open array.
+  int size = svSize(array, 1);
+  // Initialize the result value.
+  *result = 0;
+  for(size_t i = 0; i < size; ++i) {
+    svBitVecVal vec;
+    svGetBitArrElemVecVal(&vec, array, i);
+    *result += vec;
+  }
+}
+```
+
+```scala
+object Sum extends DPINonVoidFunctionImport[UInt] {
+  override val functionName = "sum"
+  override val ret = UInt(32.W)
+  override val clocked = false
+  override val inputNames = Some(Seq("array"))
+  override val outputName = Some("result")
+  final def apply(array: Vec[UInt]): UInt = super.call(array)
+}
+
+val io = IO(new Bundle {
+  val a = Input(Vec(3, UInt(32.W)))
+  val b = Input(Vec(6, UInt(32.W)))
+})
+
+val sum_a = Sum(io.a) // compute a[0] + a[1] + a[2]
+val sum_b = Sum(io.b) // compute b[0] + ... + b[5]
+```
+
+# FAQ
+
+* Can we export functions? -- No, not currently. Consider using a black box for such functionality.
+* Can we call a DPI function in initial block? -- No, not currently. Consider using a black box for initialization.
+* Can we call two clocked DPI calls and pass the result to another within the same clock? -- No, not currently. Please merge the DPI functions into a single function.
```

Comment:
I'm not sure what this last point means.

It feels like what we actually want is a light explanation of the underlying "framework" for the DPI calls, and that the above comment is actually a restriction imposed on this framework. 

For example,  is this true?: "All DPI calls on the same clock are dispatched at the same time. Because of this, it is not possible to use the result of one DPI call as an argument to another."

Reply:
> All DPI calls on the same clock are dispatched at the same time. Because of this, it is not possible to use the result of one DPI call as an argument to another

Yes, that's correct. This is fundamental restrictions for side-effecting operations in Chisel but DPI is a first example that has both side-effect and results.